src/coffea/nanoevents/methods/vector.py

```diff
@@ -1061,6 +912,52 @@ def negative(self):
         )
 
 
+_binary_dispatch_cls = {
+    "TwoVector": TwoVector,
+    "PolarTwoVector": TwoVector,
+    "ThreeVector": ThreeVector,
+    "SphericalThreeVector": ThreeVector,
+    "LorentzVector": LorentzVector,
+    "PtEtaPhiMLorentzVector": LorentzVector,
+    "PtEtaPhiELorentzVector": LorentzVector,
+}
+_rank = [TwoVector, ThreeVector, LorentzVector]
+
+for lhs, lhs_to in _binary_dispatch_cls.items():
+    for rhs, rhs_to in _binary_dispatch_cls.items():
+        out_to = min(lhs_to, rhs_to, key=_rank.index)
+        behavior[(numpy.add, lhs, rhs)] = out_to.add
+        behavior[(numpy.subtract, lhs, rhs)] = out_to.subtract
+
+
+TwoVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
+TwoVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
+TwoVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821
+
+PolarTwoVectorArray.ProjectionClass2D = PolarTwoVectorArray  # noqa: F821
+PolarTwoVectorArray.ProjectionClass3D = SphericalThreeVectorArray  # noqa: F821
+PolarTwoVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821
+
+ThreeVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
+ThreeVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
+ThreeVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821
+
+SphericalThreeVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
+SphericalThreeVectorArray.ProjectionClass3D = SphericalThreeVectorArray  # noqa: F821
+SphericalThreeVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821
+
+LorentzVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
+LorentzVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
+LorentzVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821
+
+PtEtaPhiMLorentzVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
+PtEtaPhiMLorentzVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
+PtEtaPhiMLorentzVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821
+
+PtEtaPhiELorentzVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
+PtEtaPhiELorentzVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
+PtEtaPhiELorentzVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821
```

Comment:
Should this be? -
```suggestion
TwoVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
TwoVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
TwoVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821

PolarTwoVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
PolarTwoVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
PolarTwoVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821

ThreeVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
ThreeVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
ThreeVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821

SphericalThreeVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
SphericalThreeVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
SphericalThreeVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821

LorentzVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
LorentzVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
LorentzVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821

PtEtaPhiMLorentzVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
PtEtaPhiMLorentzVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
PtEtaPhiMLorentzVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821

PtEtaPhiELorentzVectorArray.ProjectionClass2D = TwoVectorArray  # noqa: F821
PtEtaPhiELorentzVectorArray.ProjectionClass3D = ThreeVectorArray  # noqa: F821
PtEtaPhiELorentzVectorArray.ProjectionClass4D = LorentzVectorArray  # noqa: F821
```

Reply:
So obviously whatever solution we pick should make sure the array type matches the available coordinates.
That said, there is a choice here, whether:
- `SphericalThreeVector + ThreeVector = ThreeVector`, or
- `SphericalThreeVector + ThreeVector = SphericalThreeVector`

and similarly whether transposing the arguments maintains this casting.

Personally, and I've mentioned this to @jpivarski in the past, I would have had everything convert to cartesian on addition, even as far as to have `SphericalThreeVector + SphericalThreeVector = ThreeVector` (as coffea does now). The motivation is that adding is simplest in cartesian coordinates and I fear that doing it in other coordinates may inflate the effects of numerical imprecision in some cases. This would need to be carefully studied of course.

For now, I would advocate for add/subtract to return cartesian coordinates.

Reply:
In Vector, longitudinal and temporal coordinates each convert to Cartesian (`z` and `t`, respectively); it's only the polar azimuthal coordinates that stay polar if both addends are already polar. That's [this function](https://github.com/scikit-hep/vector/blob/e4c3f536d8d3833ba8b31453dc7313b5b3cc5f56/src/vector/_compute/planar/add.py#L46-L54):

```python
# specialized
def rhophi_rhophi(lib, rho1, phi1, rho2, phi2):
    diff = phi2 - phi1
    u = rho2 * lib.cos(diff)
    v = rho2 * lib.sin(diff)
    return (
        lib.sqrt((rho1 + u) ** 2 + v**2),
        rectify(lib, phi1 + lib.arctan2(v, rho1 + u)),
    )
```

If this is bad (no study was performed—what metric would define "goodness"?[^1]), then it can be replaced by

```python
def rhophi_rhophi(lib, rho1, phi1, rho2, phi2):
    return xy_xy(
        lib,
        x.rhophi(lib, rho1, phi1),
        y.rhophi(lib, rho1, phi1),
        x.rhophi(lib, rho2, phi2),
        y.rhophi(lib, rho2, phi2),
    )
```

and similarly for [subtract](https://github.com/scikit-hep/vector/blob/e4c3f536d8d3833ba8b31453dc7313b5b3cc5f56/src/vector/_compute/planar/subtract.py#L46-L54).

[^1]: Arguably, it could just be the count of transcendental functions. The `x.rhophi` and `y.rhophi` functions each have one `sin` and `cos`, so that's four, and the specialized function has a `cos`, `sin`, `sqrt`, and `arctan2`, but also the `rectify`, which uses `%` to keep the phi from wrapping around. I bet the conversion to Cartesian is faster for all inputs. To quantify round-off error, it might depend on what the inputs are (a 4D space of two pairs of azimuthal coordinates).

Reply:
Thanks @nsmith-  and @jpivarski! I'd be happy to change the way things are handled in vector or keep them as it is, but in any case, I think we should get rid off `add`, `subtract`, and other methods in coffea's vector classes. Please let me know if vector should always convert azimuthal coordinates to cartesian.

Reply:
Can you try an experiment? Sample randomly two Lorentz vectors in cartesian space (set `t^2=x^2+y^2+z^2+m^2` for some randomly sampled m), then compare the result of: adding them; converting them to azimuthal coordinates and adding them? The sampling distribution could just be gaussian with a large (~50) width.

Reply:
Accuracy and/or speed?

For accuracy, note that among Python, NumPy, and Awkward Array, there are either 2 or 3 summation algorithms that differ in the least significant bits, so choice of backend matters.

For speed, naturally big arrays are more representative.

Reply:
Accuracy with NumPy:

```python
>>> def rand(n):
...     return np.random.normal(0, 1, n)
... 
>>> cartesian_a = vector.array({"x": rand(1000000), "y": rand(1000000)})
>>> cartesian_b = vector.array({"x": rand(1000000), "y": rand(1000000)})
>>> cartesian_sum = cartesian_a + cartesian_b
>>>
>>> polar_a = vector.array({"rho": cartesian_a.rho, "phi": cartesian_a.phi})
>>> polar_b = vector.array({"rho": cartesian_b.rho, "phi": cartesian_b.phi})
>>> polar_sum = polar_a + polar_b
>>>
>>> abs(cartesian_sum - polar_sum).min()
0.0
>>> abs(cartesian_sum - polar_sum).mean()
4.484766197894659e-16
>>> abs(cartesian_sum - polar_sum).max()
5.870304242705515e-15
```

Accuracy with Awkward:

```python
>>> cartesian_a = ak.zip({"x": rand(1000000), "y": rand(1000000)}, with_name="Vector2D")
>>> cartesian_b = ak.zip({"x": rand(1000000), "y": rand(1000000)}, with_name="Vector2D")
>>> cartesian_sum = cartesian_a + cartesian_b
>>>
>>> polar_a = ak.zip({"rho": cartesian_a.rho, "phi": cartesian_a.phi}, with_name="Vector2D")
>>> polar_b = ak.zip({"rho": cartesian_b.rho, "phi": cartesian_b.phi}, with_name="Vector2D")
>>> polar_sum = polar_a + polar_b
>>>
>>> ak.min(abs(cartesian_sum - polar_sum))
0.0
>>> ak.mean(abs(cartesian_sum - polar_sum))
4.488135403509464e-16
>>> ak.max(abs(cartesian_sum - polar_sum))
7.38971534463907e-15
```

Accuracy with Python:

```python
>>> cartesian_a = [vector.obj(x=x, y=y) for x, y in zip(rand(1000000), rand(1000000))]
>>> cartesian_b = [vector.obj(x=x, y=y) for x, y in zip(rand(1000000), rand(1000000))]
>>> cartesian_sum = [a + b for a, b in zip(cartesian_a, cartesian_b)]
>>>
>>> polar_a = [vector.obj(rho=a.rho, phi=a.phi) for a in cartesian_a]
>>> polar_b = [vector.obj(rho=b.rho, phi=b.phi) for b in cartesian_b]
>>> polar_sum = [a + b for a, b in zip(polar_a, polar_b)]
>>>
>>> tmp = [abs(car - pol) for car, pol in zip(cartesian_sum, polar_sum)]
>>> np.min(tmp)
0.0
>>> np.mean(tmp)
4.479177725857644e-16
>>> np.max(tmp)
6.252832847199451e-15
```

Although there were a million elementwise `a`, `b` pairs, these accuracy measurements represent 1 numerical step. The error would grow if this is compounded in a loop. (But when are you going to have coordinates changing between Cartesian and polar in a loop? Once some operation puts it in Cartesian, it will remain in Cartesian; that's an absorbing state.)

Reply:
As for time, performing the calculation in polar is only 2% slower than converting to Cartesian, then adding.

```python
>>> import numpy as np
>>> import vector
>>> import time
>>> source_ax = np.random.normal(0, 1, 100000000)
>>> source_ay = np.random.normal(0, 1, 100000000)
>>> source_bx = np.random.normal(0, 1, 100000000)
>>> source_by = np.random.normal(0, 1, 100000000)
>>> 
>>> cartesian_a = vector.array({"x": source_ax, "y": source_ay})
>>> cartesian_b = vector.array({"x": source_bx, "y": source_by})
>>> polar_a = vector.array({"rho": cartesian_a.rho, "phi": cartesian_a.phi})
>>> polar_b = vector.array({"rho": cartesian_b.rho, "phi": cartesian_b.phi})
>>> 
>>> t0 = time.perf_counter(); tmp = polar_a + polar_b; time.perf_counter() - t0
6.290486244004569
>>> t0 = time.perf_counter(); tmp = polar_a + polar_b; time.perf_counter() - t0
6.331587403001322
>>> t0 = time.perf_counter(); tmp = polar_a + polar_b; time.perf_counter() - t0
6.302776351003558
>>> 
>>> t0 = time.perf_counter(); tmp = vector.array({"x": polar_a.x, "y": polar_a.y}) + vector.array({"x": polar_b.x, "y": polar_b.y}); time.perf_counter() - t0
6.182807716999378
>>> t0 = time.perf_counter(); tmp = vector.array({"x": polar_a.x, "y": polar_a.y}) + vector.array({"x": polar_b.x, "y": polar_b.y}); time.perf_counter() - t0
6.175768633998814
>>> t0 = time.perf_counter(); tmp = vector.array({"x": polar_a.x, "y": polar_a.y}) + vector.array({"x": polar_b.x, "y": polar_b.y}); time.perf_counter() - t0
6.201813601001049
```

By comparison, if it had been Cartesian to begin with, that's 8.8× faster:

```python
>>> t0 = time.perf_counter(); tmp = cartesian_a + cartesian_b; time.perf_counter() - t0
0.7100841389983543
>>> t0 = time.perf_counter(); tmp = cartesian_a + cartesian_b; time.perf_counter() - t0
0.710351111003547
>>> t0 = time.perf_counter(); tmp = cartesian_a + cartesian_b; time.perf_counter() - t0
0.6980755539989332
```

An argument could be made to convert to Cartesian before adding, not to speed up the addition, but to speed up any _subsequent_ operations. It has long-range performance repercussions, not short-range.

But leaving polar coordinates in polar was motivated by the principle of least surprise. An expert user might know that converting to Cartesian is better for their particular computation in the long run, but it's best to leave that decision to them.

Reply:
Ok, thanks for checking. I'm ok then with leaving it to the user to decide when to change coordinates. This does mean that the current coffea vector behavior of `PtEtaPhiMLorentzVector + PtEtaPhiMLorentzVector = LorentzVector`will change. We are opting in to this now.